pkg/controller/cronjob/utils.go

```diff
@@ -121,33 +139,105 @@ func getRecentUnmetScheduleTimes(sj batchv1beta1.CronJob, now time.Time) ([]time
 		return []time.Time{}, nil
 	}
 
-	for t := sched.Next(earliestTime); !t.After(now); t = sched.Next(t) {
+	// Step backward from now, until earliestTime.
+	// Record all missed schedules between now and then.
+	for t := PreviousScheduleTime(schedule, now); t.After(earliestTime); t = PreviousScheduleTime(schedule, t) {
 		starts = append(starts, t)
-		// An object might miss several starts. For example, if
-		// controller gets wedged on friday at 5:01pm when everyone has
-		// gone home, and someone comes in on tuesday AM and discovers
-		// the problem and restarts the controller, then all the hourly
-		// jobs, more than 80 of them for one hourly scheduledJob, should
-		// all start running with no further intervention (if the scheduledJob
-		// allows concurrency and late starts).
-		//
-		// However, if there is a bug somewhere, or incorrect clock
-		// on controller's server or apiservers (for setting creationTimestamp)
-		// then there could be so many missed start times (it could be off
-		// by decades or more), that it would eat up all the CPU and memory
-		// of this controller. In that case, we want to not try to list
-		// all the missed start times.
-		//
-		// I've somewhat arbitrarily picked 100, as more than 80,
-		// but less than "lots".
-		if len(starts) > 100 {
-			// We can't get the most recent times so just return an empty slice
-			return []time.Time{}, fmt.Errorf("Too many missed start time (> 100). Set or decrease .spec.startingDeadlineSeconds or check clock skew.")
+		t = t.Add(-time.Second) // Make sure we don't return this same time again
+		// We only care about knowing 0, 1, or multiple missed start times.
+		if len(starts) > 1 {
+			return starts, nil
 		}
 	}
+
 	return starts, nil
 }
 
+// PreviousScheduleTime returns the lastest time this schedule is valid, before or at the given time.
+// If no time can be found to satisfy the schedule, return the zero time.
+func PreviousScheduleTime(schedule cron.SpecSchedule, t time.Time) time.Time {
```

Comment:
This is based on the robfig/crons module's Next() function.

Reply:
Hard-coding / relying on `cron.SpecSchedule` rather than the [`Schedule`](https://godoc.org/github.com/robfig/cron#Schedule) interface won't support the entire API surface that the cron implementation this controller uses (`robfig/cron`) provides. Specifically, the [`ConstantDelaySchedule`](https://godoc.org/github.com/robfig/cron#ConstantDelaySchedule) for crons that have a spec of `@every 10m`, or variations, aren't covered. (Though technically speaking, I don't think the `@every` directive is standard per what the API reference [points to](https://en.wikipedia.org/wiki/Cron) as acceptable cron formats and is actually a leaky abstraction)

Just one argument in favor of the binary search approach, relying only on `.Next()`.